modGui/src/org/aion/gui/model/KernelConnection.java

```diff
@@ -0,0 +1,135 @@
+    @VisibleForTesting KernelConnection(IAionAPI aionApi,
+                                        CfgApi cfgApi,
+                                        EventBus eventBus,
+                                        ExecutorService executorService) {
+        this.api = aionApi;
+        this.cfgApi = cfgApi;
+        this.eventBus = eventBus;
+        this.backgroundExecutor = executorService;
+    }
+
+    /** Connect to API */
+    public void connect() {
+        if (connectionFuture != null) {
+            connectionFuture.cancel(true);
+        }
+        connectionFuture = backgroundExecutor.submit(() -> {
+            synchronized (api) {
+                LOG.trace("About to connect to API");
+                ApiMsg msg =  api.connect(getConnectionString(), true);
+                if(msg.isError()) {
+                    // since api.connect called with reconnect = true, it should
+                    // block until msg is not error so this shouldn't happen, but
+                    // log if it does.
+                    LOG.error("Error connecting to Api.  ErrorCode = {}.  ErrString = {}",
+                            msg.getErrorCode(), msg.getErrString());
+                } else {
+                    eventBus.post(new RefreshEvent(RefreshEvent.Type.OPERATION_FINISHED));
+                }
+            }
+        });
+    }
+
+    /** Disconnect from API. */
+    public void disconnect() {
+        if (!isConnected()) {
+            return;
+        }
+        if(connectionFuture != null) {
+            connectionFuture.cancel(true);
+        }
+        if(disconnectionFuture != null) {
+            disconnectionFuture.cancel(true);
+        }
+        disconnectionFuture = backgroundExecutor.submit(() -> {
+            synchronized (api) {
+                LOG.trace("About to destroy API");
+                api.destroyApi().getObject();
+            }
+        });
+
+    }
+
+    /**
+     * @return whether API is connected
+     */
+    public boolean isConnected() {
+        synchronized (api) {
+            return api.isConnected();
+        }
+    }
+
+    /**
+     * Intended to only be used by AbstractAionApiClient.
+     *
+     * @return api that the kernel is connected to
+     */
+    @VisibleForTesting IAionAPI getApi() {
+        // Impl note: Can make this public if there's a good reason for other classes
+        // to call this in the future.  Because of the non-thread-safe nature of API,
+        // currently it is restricted and the recommended way to call the API is to subclass
+        // from AbstractAionApiClient which provides subclasses with synchronized blocks to
+        // call their critical sections with.
+        return this.api;
+    }
+
+    private String getConnectionString() {
+        final String protocol = "tcp";
+        final String ip = Preconditions.checkNotNull(cfgApi.getZmq().getIp());
+        final String port = Preconditions.checkNotNull(String.valueOf(cfgApi.getZmq().getPort()));
```

Comment:
If the Preconditions.checkNotNull fails (and throws NPE), there is nothing we can really do to recover.  So that's why I did not put a catch.

I guess we can catch it and present a nicer error via GUI instead of the program just crashing and exiting, but this is something I'm planning to add later.

Let me know if you still want me to add a catch.